update to reflect php-fpm needs for system environment varibales

mmattel opened the pull request:
Related to https://github.com/owncloud/core/pull/15617 (Checks if getenv returns proper system environment variable results)

This is the needed text for the php section of the documentation

mmattel commented:
Additional note:
this PR can be merged even the core PR does not get merged, but must be merged when the referenced core PR gets merged.

carlaschroder commented:
Do you mean for this to go to master, or stable8? Master is 8.1.

mmattel commented:
At least master, valid for all versions, necessary for master if referenced PR from core gets merged.

Xenopathic commented on admin_manual/installation/source_installation.rst:
```diff
@@ -277,6 +277,15 @@ Keep in mind that changes to php.ini may have to be done on more than one ini fi
+populated in the same way compared when using php-cli. A php call like ``getenv('PATH');`` can therefore return an empty result. 
+Even fallback content for those variables is most times defined, this may not match 100% of the needs. 
+In those cases, the variables need to be set manually in the corresponding php-fpm ini file. Usually the location 
+for these ini files can be found at: ``\etc\php5\fpm\``
```
Perhaps a little bit more here explaining the required syntax? E.g.

> The PATH variable can be set by putting `env[PATH] = /usr/bin:/somewhere/something` in your PHP-FPM pool file

carlaschroder replied:
The slashes are backwards on `\etc\php5\fpm\`

carlaschroder commented:
> The PATH variable can be set by putting env[PATH] = /usr/bin:/somewhere/something in 
> your  PHP-FPM pool file

What is a pool file?

Xenopathic commented:
@carlaschroder The file that defines a 'pool', much like a virtual host for Apache.

carlaschroder commented:
@Xenopathic What file exactly do users edit? Is it something like `/etc/foo/pool`? 

This also needs a brief explanation of what PHP-FPM is for and why users would want to use it. The mention of it in this doc comes out of the blue with no context, and it's not a required PHP module, so it needs more explanation.

Xenopathic commented:
@carlaschroder Ah, PHP-FPM is used with non-Apache webservers like nginx: https://doc.owncloud.org/server/8.1/admin_manual/installation/nginx_configuration.html. In fact, I kinda think that this new documentation should go in the nginx page, where it is most relevant. That page already notes some other things that might need to be configured for PHP-FPM.

carlaschroder commented:
@Xenopathic, feel free to open a separate PR documenting PHP-FPM.

@mmattel, please make the one correction I noted, and you will backport as appropriate, correct?

thanks!

Xenopathic commented:
@carlaschroder IMHO an admin should know about PHP-FPM if they are setting up nginx - no PHP applications work without it. By that extension it is a bit redundant to document it here :smile:

carlaschroder commented:
@Xenopathic Wrong. We get users of all experience and ability levels, LAMP stacks are complex, and it is always a mistake to assume that ownCloud users and admins know everything already. And you did open that door yourself.

The purpose of a pull request is to provide specific ready-to-apply patches. I've asked you several specific questions in this PR and another one, and not received pertinent replies. I suggest that general discussions belong elsewhere.

BTW php-fpm is not just for Nginx.

Xenopathic commented:
@carlaschroder Apologies - I only see this from the perspective of an experienced admin, so I don't know what a new admin or anyone unfamiliar with PHP-FPM would think. As far as I'm aware, we only recommend use of Apache with mod_php, but I can document some more about PHP-FPM if that is required. Only issue is where to put it - in a generic place like this PR, or under the nginx page (where there are already some references to PHP-FPM) or a new page entirely.

mmattel commented:
Come on, you both have valid points and you both are well experienced.

I have changed the text according the notes you stated plus improved the content a bit.

Even if we recommend apache, we also public support nginx, therefore we should write the text in an appropriate place which is imho this one. I have looked in advance where to put it and only writing to nginx, which I evaluated, is not the right choice. Please see as example https://www.linode.com/docs/websites/apache/running-fastcgi-php-fpm-on-debian-7-with-apache
Our document clearly prefers apache, but definitely links to other webserver configuration pages too.

It is also right that at the nginx conf page, we have a statement regarding php-fpm. As the statement is true also for other webservers, I vote (and volunteer) for moving this here. Pls let me know.

At the end of the day, we want to minimize problems. We have to acknowledge that all admins have a learning curve and texts like this help making owncloud running smooth. That is the goal.

Note: I would, and this is my oponion, put the apache text component (incl enabling ssl) of this page down, just before other webservers, then things are more clear, but as said just my 2c

oparoz commented on admin_manual/installation/source_installation.rst:
```diff
@@ -277,6 +277,32 @@ Keep in mind that changes to php.ini may have to be done on more than one ini fi
+**Notes when using php-fpm:**
+
+It may be the case when using ``php-fpm``, system environment variables like PATH, TMP or others are not automatically 
+populated in the same way compared when using ``php-cli``.
```
I think the line break is unwanted

oparoz commented on admin_manual/installation/source_installation.rst:
```diff
@@ -277,6 +277,32 @@ Keep in mind that changes to php.ini may have to be done on more than one ini fi
+populated in the same way compared when using ``php-cli``.
+
+A php call like ``getenv('PATH');`` can therefore return an empty result.
+Even fallback content for those variables is most times defined, this may not match 100% of the needs. 
```
I don't understand this sentence

mmattel commented:
Info: the core PR referenced has been merged. We should do final corrections if necessary and merge this also.

oparoz commented on admin_manual/installation/source_installation.rst:
```diff
@@ -277,6 +277,32 @@ Keep in mind that changes to php.ini may have to be done on more than one ini fi
+	;env[TEMP] = /tmp
+
+You can either just uncomment the existing one and/or copy the output from the command shell eg: ``printenv PATH`` to 
+the environment variable to be set. If no system environment variable is present to be uncommented, 
```
line break

oparoz commented on admin_manual/installation/source_installation.rst:
```diff
@@ -277,6 +277,32 @@ Keep in mind that changes to php.ini may have to be done on more than one ini fi
+
+Please keep in mind, that it is possible to create different settings for ``php-cli`` and ``php-fpm``
+and even for different domains/websites. Please take care to take the right one. Your result can be tested with ``phpinfo()``
+which must accessed via a browser to your domain/website.
```
line break

oparoz replied:
> Please take care to take the right one
You should make sure you modify the one which applies to your ownCloud installation

mmattel replied:
Line breakes are automatically corrected. That´s just a presentation thing of the raw data, final is fine. I checked it in my browser. Please correct me if you experience differences when you test this PR.

Nobody can say in advance which one applies to owncloud. This is the last step the admin has to decide on his own. This is also the reason why I added the examples with ``;env[]`` so the admin can search for.

oparoz replied:
Well, the markdown parser won't create a new line unless you leave a space in between paragraphs, but it doesn't mean that it's wrong in the original text. I'll let @carlaschroder decide what's right here.

And an admin should know which files he needs to modify in order to configure PHP for his domain or he's in big trouble when something goes wrong. 
Maybe something like this:
`, so make sure you modify the files which apply to the PHP-FPM pool your ownCloud installation belongs to.`

oparoz commented:
I also vote to move everything related to PHP-FPM to a specific page since this is totally independent of the web server and requires some attention when configuring.
I wouldn't give a course on PHP-FPM there, since the doc is not giving one on PHP or how to install its modules, but a short introduction as to why it's the preferred way to use PHP, especially for a system like ownCloud, would be a nice touch.

carlaschroder commented:
Re: Nginx-- we do not offically support Nginx. However, it's a popular and good Web server so we maintain docs for it.

Re: the PHP-FPM pool file-- is that the same as `/etc/php5/fpm/` ? Computers are exactly literal and specific, so our documentation must also be exactly literal and specific. For example, in older ownCloud docs I've seen "the configuration file" used instead of "config.php". Little things like that create confusion and frustrated users; they shouldn't have to guess the meaning of vague terms like "the configuration file" or "the pool file" or whatever it might be.

The focus of the ownCloud manuals is configurations that are pertinent to running an ownCloud server, without getting bogged down in re-documenting every aspect of LAMP stacks and system administration. So, in the case of PHP-FPM,  any config items that are directly necessary for ownCloud are useful. We don't need a complete course on PHP-FM. I see that this PR has some new examples, yay, so I'll look at today.

oparoz commented:
Regarding the location of files, there is no single answer. In a Directadmin environment, it could be 
/usr/local/directadmin/data/users/<user>/php/php-fpm55.conf or even /usr/local/directadmin/data/users/<user>/php/php-fpm55.conf.custom1 or /usr/local/directadmin/data/users/<user>/php/php-fpm55.conf.custom2 and that's just for one type of environment and for one pool. A sysadmin may want to adjust the global PHP-FPM settings and that file would be located in the folder where PHP is installed.

That's the reason I think we can't give an exact location and should trust that sys admins know what they're doing.
Since the same page has an Ubuntu example, maybe the location for Ubuntu shoud be added, so that those people with little experience and who are trying to follow the guide aren't completely lost when told to modify their configuration.

mmattel commented:
@oparoz 
> Regarding the location of files, there is no single answer. 

That was the reason I originally had in the core PR a middle line where the path used by php-fpm was printed. But my idea was rejected an I was asked to taked that line out. Now the PR is merged...
(You gave the best examples above how comlex this can become...)

> Since the same page has an Ubuntu example, maybe the location for Ubuntu shoud be added

There is already a text added in this PR: 
--For Ubuntu, the location for these ini files can be found at: ``/etc/php5/fpm/``--
But I can write ini/config files to make it more clear

@carlaschroder 
The Ubuntu example path is a starting point. Under this directory, you can have several ini files and subdirectories with additional ini/config files. One, and commonly used subdirectory there is called ``pool.d`` where there is a file called ``www.conf``. This file in Ubuntu contains the commented out example parameters. But remember, this is not a must! I used the term pool to give some help, for people who want to look in the internet for more details.
One of the problem points we ran into was a not easy to find issue related to an incomplete php-fpm configuration. We fixed that on the core side but it additional needs some documentation, that´s this PR. And of course, it is not intended to do a complete php-fpm config and install guide here, just that where we see issues owncloud related - at least my understanding.

@carlaschroder 
Question: what about the idea to move the apache part into a seperate document like nginx?

@carlaschroder 
Question: shall I mirgate the small php-fpm part of nginx to this PR to have it at one place?

carlaschroder commented:
I've been reading up on PHP-FPM, and reviewing https://github.com/owncloud/core/pull/15617. Thanks for taking on the documentation part, @mmattel. I think I can help simplify this.  As you  noted, we don't want to deliver a comprehensive PHP-FPM course.  CentOS/Red Hat/Fedora has a default /etc/php-fpm.d/www.conf, and  Ubuntu/Mint has /etc/php5/fpm/pool.d/www.conf. Those two examples cover the great majority of Linux installations. So I suggest writing up the simplest possible example configuration that covers the most important items, referencing those two www.conf files. Put it on a separate page, and then I can link to it from the Apache and Nginx pages. 

Over time I, or whoever feels motivated, can add to the PHP-FPM material. But for now I agree that keeping this relevant to https://github.com/owncloud/core/pull/15617 is good enough. Does that work for you, @mmattel?

carlaschroder commented:
No wait :) Will a single example in source_installation.rst work for both Apache and Nginx?

carlaschroder commented:
Sorry for making this over-complicated-- I should have read up on PHP-FPM first.

oparoz commented:
@mmattel - The Ubuntu reference is fine. I failed to re-read the whole thing when I was making the comment about that one sentence.

And I think there should be a note that if people are using shared hosting or a control panel to manage their VM or server, then they should check the documentation which comes with that software as the configuration files are almost certain to be located somewhere else, for security and flexibility reasons.

mmattel commented:
@carlaschroder 
I updated and cleaned the texts and squashed the commits.

What about the two questions I asked. I could easily add the first one (or we do this one and create another PR for that move)... and the second one would make things imho more consistent (but may be blasphemic to others :smile: )

carlaschroder commented:
@mmattel It looks good, thanks.

> Question: what about the idea to move the apache part into a seperate document like nginx?
Leave it where it is for now. The page covers a typical LAMP installation, so moving the Apache section means reworking the whole page.

@carlaschroder
> Question: shall I mirgate the small php-fpm part of nginx to this PR to have it at one place?
Yes, that makes sense, and replace it with a cross-reference. If you're not familiar with cross-references it looks like this-- in  source_installation.rst make a label and a subheading (it has to be a heading or a subhead):

```
.. _using_php-fpm:
Notes when using php-fpm
------------------------
```

Then you can use the label (without the leading underscore) on the Nginx page to create a link to that section:

```
:ref:`using_php-fpm`
```

Again, my apologies for creating confusion. Some days are good days, some are not!

mmattel commented:
@carlaschroder 
updated texts, added reference as suggested, integrated php-fpm text from nginx, removed the php-fpm text from nginx, squashed. Please check if everything is fine now.

Your confusion is welcomed, it gives the chance to clarify things. Look at the result :smile:

oparoz commented:
@mmattel - Good job. I think users should be able to configure their environment properly using these instructions.

I just have one comment. I don't want to hurt any feelings, but a few sentences need to be converted into "proper" English. I don't know if @carlaschroder is responsible for that or not, but on a couple of occasions, it's really difficult to understand what the authors mean.

mmattel commented:
@oparoz, @carlaschroder 
No worries, feel free to correct / improve my writing. I am not a native speaker and I am happy to learn...
Let me know what needs to be done and I will implement it.
Note: I can do the changes earliest on monday as I am offsite over the weekend

carlaschroder commented:
@oparoz, yes I am responsible for translating developer -> sysadmin :)
I'm going to go ahead and merge, and then copyedit. Thanks again, this good work.

carlaschroder commented:
https://github.com/owncloud/documentation/pull/1085

mdik commented:
Is there a security risk if PATH is not set?

I don't understand the implications of this :-(

settermjd commented:
@PVince81 @DeepDiver1975 - your thoughts, please.

PVince81 commented:
@phisch @IljaN @butonic
